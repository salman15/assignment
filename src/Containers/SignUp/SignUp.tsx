import React, { useEffect, useState } from "react";
import Button from "../../Components/Button/Button";
import Form from "../../Components/Form/Form";
import Input from "../../Components/Form/Input";
import { Column, Container, WarningText } from "../../Components/Layout/Layout";
import { UsePost, UseGet } from "../../Hooks";
import {
  validateSignUp,
  validEmail,
  validName,
  validPassword,
} from "../../Utils/Validate";

const SignUp = () => {
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [showValidationErrors, setShowValidationErrors] = useState(false);
  const [newAccount, setNewAccount] = useState(false);
  const { getLoading, getError, getData } = UseGet("delay/4");
  const { postResponse, postLoading, postError, sendPost } = UsePost("post", {
    firstName: firstName,
    lastName: lastName,
    email: email,
    password: password,
  });

  useEffect(() => {
    if (postResponse.status >= 200 && postResponse.status < 300) {
      setNewAccount(true);
    }
  }, [postResponse]);

  /**
   *
   */
  const handleSubmit = (event: any) => {
    event.preventDefault();
    if (
      validateSignUp({
        firstName: firstName,
        lastName: lastName,
        email: email,
        password: password,
      })
    ) {
      getData(true);
      sendPost(true);
    } else {
      setShowValidationErrors(true);
    }
  };

  return (
    <Container>
      {newAccount ? (
        <Column>
          <h1>Account created</h1>
          <Button action={() => setNewAccount(false)}>Go back</Button>
        </Column>
      ) : (
        <Form
          title="Create an new account"
          description="everything with a * is required"
          error={getError || postError}
          submit={handleSubmit}
        >
          {(getError || postError) && (
            <WarningText>Something went wrong</WarningText>
          )}

          <Input
            label="First name"
            warning={showValidationErrors && !validName(firstName)}
            warningText="Incorrect name"
            placeholder="enter your name"
            inputValue={firstName}
            setInputValue={setFirstName}
            required
            type="first-name"
          />
          <Input
            label="Last name"
            type="last-name"
            warning={showValidationErrors && !validName(lastName)}
            warningText="Incorrect last name"
            placeholder="enter your last name"
            inputValue={lastName}
            setInputValue={setLastName}
            required
          />
          <Input
            label="Email"
            type="email"
            warning={showValidationErrors && !validEmail(email)}
            warningText="Incorrect email"
            placeholder="enter your email"
            inputValue={email}
            setInputValue={setEmail}
            required
          />
          <Input
            label="Password"
            warning={showValidationErrors && !validPassword(password)}
            warningText="Password must contain at least one lowercase letter and one uppercase letter"
            placeholder="enter your password"
            inputValue={password}
            setInputValue={setPassword}
            required
            type="password"
          />
          <Button
            disabled={
              firstName.length < 1 ||
              lastName.length < 1 ||
              email.length < 1 ||
              password.length < 8 ||
              getLoading ||
              postLoading
            }
          >
            {getLoading || postLoading ? "loading" : "Create account"}
          </Button>
          <p>
            Already have an account?{" "}
            <Button type="link" link="#">
              Sign in
            </Button>
          </p>
        </Form>
      )}
    </Container>
  );
};

export default SignUp;
