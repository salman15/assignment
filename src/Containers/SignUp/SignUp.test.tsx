import React from "react";
import { render, fireEvent } from "@testing-library/react";
import SignUp from "./SignUp";

describe("Test the SignUp component", () => {
  it("Test if SignUp component renders with children and input values can be adjusted", () => {
    // Act
    const { container, getByLabelText } = render(<SignUp />);

    const emailInput = getByLabelText(`email-input`);
    const firstNameInput = getByLabelText(`first-name-input`);
    const LastNameInput = getByLabelText(`last-name-input`);
    const passwordInput = getByLabelText(`password-input`);

    fireEvent.change(emailInput, { target: { value: "john@email.com" } });
    fireEvent.change(firstNameInput, { target: { value: "John" } });
    fireEvent.change(LastNameInput, { target: { value: "Wick" } });
    fireEvent.change(passwordInput, { target: { value: "ImJohnWick" } });
    // Assert
    expect(container).toMatchSnapshot();

    expect(emailInput.value).toBe("john@email.com");
    expect(firstNameInput.value).toBe("John");
    expect(LastNameInput.value).toBe("Wick");
    expect(passwordInput.value).toBe("ImJohnWick");
  });
});
