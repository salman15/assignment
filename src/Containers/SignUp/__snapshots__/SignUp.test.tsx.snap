// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the SignUp component Test if SignUp component renders with children and input values can be adjusted 1`] = `
<div>
  <div
    class="sc-pFZIQ cyLYfa"
  >
    <form
      class="sc-hKgILt iuDjhV"
    >
      <h1>
        Create an new account
      </h1>
      <p>
        everything with a * is required
      </p>
      <div
        class="sc-fubCfw elfxoX"
      >
        <p
          class="sc-jSgupP eHZQIR"
        >
          First name
           
          <span
            class="sc-gKsewC eOGKIx"
          >
            *
          </span>
        </p>
        <input
          aria-label="first-name-input"
          autocomplete="first-name"
          class="sc-eCssSg crOsbw"
          name="label"
          placeholder="enter your name"
          type="first-name"
          value="John"
        />
      </div>
      <div
        class="sc-fubCfw elfxoX"
      >
        <p
          class="sc-jSgupP eHZQIR"
        >
          Last name
           
          <span
            class="sc-gKsewC eOGKIx"
          >
            *
          </span>
        </p>
        <input
          aria-label="last-name-input"
          autocomplete="last-name"
          class="sc-eCssSg crOsbw"
          name="label"
          placeholder="enter your last name"
          type="last-name"
          value="Wick"
        />
      </div>
      <div
        class="sc-fubCfw elfxoX"
      >
        <p
          class="sc-jSgupP eHZQIR"
        >
          Email
           
          <span
            class="sc-gKsewC eOGKIx"
          >
            *
          </span>
        </p>
        <input
          aria-label="email-input"
          autocomplete="email"
          class="sc-eCssSg crOsbw"
          name="label"
          placeholder="enter your email"
          type="email"
          value="john@email.com"
        />
      </div>
      <div
        class="sc-fubCfw elfxoX"
      >
        <p
          class="sc-jSgupP eHZQIR"
        >
          Password
           
          <span
            class="sc-gKsewC eOGKIx"
          >
            *
          </span>
        </p>
        <input
          aria-label="password-input"
          autocomplete="password"
          class="sc-eCssSg crOsbw"
          name="label"
          placeholder="enter your password"
          type="password"
          value="ImJohnWick"
        />
      </div>
      <button
        class="sc-bdfBwQ sc-gsTCUz iDdfZq bhjXvn"
        href=""
      >
        Create account
      </button>
      <p>
        Already have an account?
         
        <a
          class="sc-bdfBwQ sc-dlfnbm iDdfZq bgwCyg"
          href="#"
        >
          Sign in
        </a>
      </p>
    </form>
  </div>
</div>
`;
